File: python/단지번호붙이기(2667) copy.py
from collections import deque

n = int(input())
graph = [
    list(map(int, input().strip()))
    for _ in range(n)
]

visited = [[False] * n for _ in range(n)]

# 상, 하, 좌, 우
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

cnt_list = []

def bfs(x,y):
    queue = deque()
    queue.append((x,y))
    visited[x][y] = True
    cnt = 1

    while queue:
        x,y = queue.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if nx >=0 and nx < n and ny >=0 and ny < n and graph[nx][ny] == 1 and not visited[nx][ny]:
                cnt += 1
                visited[nx][ny] = True
                queue.append((nx,ny))
    cnt_list.append(cnt)

total_cnt = 0
for i in range(n):
    for j in range(n):
        if graph[i][j] == 1 and not visited[i][j]:
            bfs(i,j)
            total_cnt += 1

print(total_cnt)
cnt_list.sort()
for elem in cnt_list:
    print(elem, end='\n')

File: js/단지번호붙이기(2667) copy.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

const n = Number(input[0]);
const graph = input.slice(1, n+1).map((line) => line.split('').map(Number));
const visited = Array.from({length:n}, () => Array(n).fill(false));

// 상, 하, 좌, 우
const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

let cnt_list = [];

function bfs(x,y) {
    queue = [];
    queue.push([x,y]);
    visited[x][y] = true;
    let cnt = 1;

    while (queue.length) {
        const [x,y] = queue.shift();
        for (let i=0; i<4; i++) {
            let nx = x + dx[i];
            let ny = y + dy[i];

            if (nx >=0 && nx < n && ny >=0 && ny < n && !visited[nx][ny] && graph[nx][ny] === 1) {
                cnt++;
                visited[nx][ny] = true;
                queue.push([nx,ny]);
            }
        }
    }
    cnt_list.push(cnt);
}

let total_cnt = 0;

for (let i=0; i<n; i++) {
    for (let j=0; j<n; j++) {
        if (!visited[i][j] && graph[i][j] === 1) {
            bfs(i,j);
            total_cnt++;
        }
    }
}

cnt_list.sort((a,b) => a-b);
console.log(total_cnt);
for (let elem of cnt_list) {
    console.log(elem);
}